fixes #19995 - search now uses turbolinks

ohadlevy opened the pull request:
this avoids full page reloads when using search.

this also adds the urijs package, and while it can be done
without it, I prefered using it in order to be able to
support later on the per_page feature.

ohadlevy commented:
before:
![before](https://user-images.githubusercontent.com/20167/27084659-5a35aef6-5055-11e7-9849-ce25b8627dce.gif)

after:
![after](https://user-images.githubusercontent.com/20167/27084664-5ebff4b8-5055-11e7-836a-3daeb459f81a.gif)

ohadlevy commented:
@tbrisker, @matanwerbner  any idea how should i test this? thanks.

ohadlevy commented:
test failures seems unrelated.

ohadlevy commented:
added tests.

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -80,3 +81,22 @@ export function initTypeAheadSelect(input) {
+  let uri = new URI(window.location.href);
+  let search;
+
+  if (element !== undefined) {
```
Is there any place the search field doesn't have id `#search` or isn't the child `.autocomplete-input`? I would rather use a sane default then a condition.

ohadlevy replied:
One of the reasons we can't have multiple implementation of search in the same page is because of the ID selector (see roles and filters). I would like to decouple ourself from that pattern and allow using multiple search fields when appropriate.

boaz0 reviewed (approved):
LGTM

boaz0 commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -80,3 +81,22 @@ export function initTypeAheadSelect(input) {
+  let uri = new URI(window.location.href);
+  let search;
+
+  if (element !== undefined) {
```
maybe you can use a ternary operator here:
```
search = (element !== undefined) ? $(element).find('.autocomplete-input') : $('#search')
```
Nevertheless, I am not sure how readable will it be.

ohadlevy replied:
thanks! I wrote it like that as I know https://github.com/theforeman/foreman/pull/4550/files#diff-8bb38b9ee188f2de0368a1fcad1ede41R87 (once it will be merged) will probably use this code path too..

tbrisker commented:
Looks good and works well, just ~one~ a few inline comment.

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -80,3 +81,22 @@ export function initTypeAheadSelect(input) {
+    search = $('#search');
+  }
+
+  if (search[0] !== undefined) {
```
can this condition actually occur? that means there is no search element present on the page

ohadlevy replied:
This is maybe premature, but the idea behind this function is to be called when table related data has changed, including per_page pagination change. in rare cases where there is no search for a given table that can happen.

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -80,3 +81,22 @@ export function initTypeAheadSelect(input) {
 }
+
+// send get forms through turbolinks
+export function tableUpdated(element) {
```
maybe name this more descriptively, something like searchTable or something? the name is a bit confusing

ohadlevy replied:
see my previous comment about this? how would you call it?

ohadlevy commented:
@tbrisker rebased and addressed your comments.

ohadlevy committed:
fixes #19995 - search now uses turbolinks

this avoids full page reloads when using search.

this also adds the urijs package, and while it can be done
without it, I prefered using it in order to be able to
support later on the per_page feature.

ohadlevy commented:
@tbrisker test failure seems unrelated. @dLobatog another one for your list?

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -80,3 +81,19 @@ export function initTypeAheadSelect(input) {
+    if (search !== undefined) {
+      uri.setSearch('search', search.trim());
+    }
+  /* eslint-disable no-undef */
```
nitpick: should be on next line

tbrisker reviewed (changes requested):
just a tiny nitpick and I'll merge.

ohadlevy commented:
@tbrisker updated.

tbrisker commented:
Thanks @ohadlevy !

tbrisker commented:
Please make sure to add uri.js to https://github.com/theforeman/foreman-packaging/tree/rpm/develop
